File: src/ai/schemas.py
"""Pydantic models for Gemini JSON outputs."""

from __future__ import annotations

from typing import Literal, Optional

from pydantic import BaseModel, Field


class CategorizationItem(BaseModel):
    index: int
    description: str
    amount: float
    category: Literal["necessario", "superfluo"]


class CategorizationResponse(BaseModel):
    reference_month: Optional[str] = None
    items: list[CategorizationItem] = Field(default_factory=list)


class MonthOverMonthItem(BaseModel):
    metric: Literal["total", "necessario", "superfluo"]
    direction: Literal["up", "down", "flat"]
    comment: str


class ConsultingResponse(BaseModel):
    summary: str
    month_over_month: list[MonthOverMonthItem] = Field(default_factory=list)
    leaks: list[str] = Field(default_factory=list)
    tips: list[str] = Field(default_factory=list)

File: src/lib/ai/schemas.ts
import { z } from "zod";

export const extractionResponseSchema = z.object({
  reference_month: z.string().regex(/^\d{4}-\d{2}$/),
  transactions: z.array(
    z.object({
      date: z.string().regex(/^\d{4}-\d{2}-\d{2}$/),
      description: z.string().min(1),
      amount: z.number().positive(),
    })
  ),
});

export type ExtractionResponse = z.infer<typeof extractionResponseSchema>;

export const categorizationResponseSchema = z.object({
  items: z.array(
    z.object({
      index: z.number().int().min(0),
      category: z.enum(["necessario", "superfluo", "investimento"]),
    })
  ),
});

export type CategorizationResponse = z.infer<
  typeof categorizationResponseSchema
>;

export const consultingResponseSchema = z.object({
  summary: z.string(),
  month_over_month: z.array(
    z.object({
      metric: z.string(),
      direction: z.enum(["up", "down", "flat"]),
      comment: z.string(),
    })
  ),
  leaks: z.array(z.string()),
  tips: z.array(z.string()),
});

export type ConsultingResponse = z.infer<typeof consultingResponseSchema>;
